### DISCUSSION TITLE
pypdfium2 can now directly render to numpy arrays

### DISCUSSION THREAD

**Discussion by mara004** [mindee/doctr#1032] [Show and tell]
Hello, pypdfium2 author here again.

I'm happy to announce that pypdfium2 can now directly render to numpy arrays. without the necessity of going through an intermediate `PIL.Image.Image` object (commit https://github.com/pypdfium2-team/pypdfium2/commit/7c4285ca1bfd599e2bd63e520c90339a2b8ead30).
I also integrated pdfium's option to reverse the byte order so that you can get colour format `RGB(A)` rather than `BGR(A)` (commit https://github.com/pypdfium2-team/pypdfium2/commit/5d858f2b846b3808aebb6588993e231fc52c6635).

The `io/pdf.py` module could basically be changed like this to use `render_tonumpy()` rather than `render_topil()`:
```python
    # Rasterise pages to numpy arrays with pypdfium2
    with pdfium.PdfDocument(file, password=password) as pdf:
        return [a for a, _ in pdf.render_tonumpy(scale=scale, rev_byteorder=True, **kwargs)]
```

The new features are not released yet, but you can try them by installing pypdfium2 from its main branch:
```bash
python3 -m pip install git+https://github.com/pypdfium2-team/pypdfium2.git@main
```
I plan to increment major version with the next release because of some API-breaking changes,[^1] so it might yet take some time.

[^1]: (none of which impact `doctr`, though)

**Comment by frgfm** [comment]
Hello @mara004 :wave:

Always glad to hear about the new features you've been working on :)
I suggest that we wait for the feature to make its way to a stable release before we integrate it here, if that's alright with you!

Best :v:

**Comment by mara004** [comment]
For reference, https://github.com/mindee/doctr/pull/1040 is related to this thread.

**Comment by mara004** [comment]
I've written a small script to benchmark the various ways how we can render to numpy arrays using pypdfium2, to check that the new functions are actually faster than using a PIL image as intermediary, and to investigate what impact byte order reversion has.
The test case `render_numpy_reverse_pdfium()` is using the strategy I would suggest for doctr.
```python
import time
import numpy
import pypdfium2 as pdfium


pdf = pdfium.PdfDocument("/path/to/some_file.pdf")  # enter file path here
kwargs = dict(
    scale = 3,
    # page_indices = [i for i in range(25)],  # limit number of pages, if necessary
)


# render to numpy array using pypdfium2's helper function
def render_numpy():  # BGR
    images = list(pdf.render_tonumpy(**kwargs))


# render to numpy array using pypdfium2's helper function, reversing byte order with pdfium
def render_numpy_reverse_pdfium():  # RGB
    images = list(pdf.render_tonumpy(**kwargs, rev_byteorder=True))


# render to numpy array using a PIL image as intermediary, reversing byte order with PIL
def render_pil_numpy_reverse_pil():  # RGB
    images = [numpy.asarray(img) for img in pdf.render_topil(**kwargs)]


# render to numpy array using a PIL image as intermediary, reversing byte order with pdfium
def render_pil_numpy_reverse_pdfium():  # RGB
    images = [numpy.asarray(img) for img in pdf.render_topil(**kwargs, rev_byteorder=True)]


functions = (
    render_numpy,
    render_numpy_reverse_pdfium,
    render_pil_numpy_reverse_pil,
    render_pil_numpy_reverse_pdfium,
)

def main():
    for func in functions:
        start = time.time()
        print("Running %s ..." % func.__name__)
        func()
        print("-> %s seconds" % (time.time() - start), end=2*"\n")

if __name__ == "__main__":
    main()
```

On my test document, the new `render_tonumpy()` function is distinctly faster than rendering to a PIL image and converting that. The performance impact of pdfium byte order reversion seems to vary depending on used resolution, sometimes for better and sometimes for worse.

**Reply by felixdittrich92** [reply]
Thanks a lot @mara004 🙌 @frgfm wdyt ?

**Reply by frgfm** [reply]
Avoiding intermediate format is always a good thing, especially because we're trying to get numpy arrays.
Let's switch to this (after running a few latency & integrity checks) once this feature made its way to a stable release on your library @mara004 :)

**Comment by mara004** [comment]
Another note: I've added a generic `render_to()` strategy that allows callers to plug in any custom converter when using the multipage renderer. Target formats like `PIL.Image.Image` and `numpy.ndarray` are now implemented with built-in converters.
While `render_topil()` etc. are retained as aliases for backwards compatibility, I'd recommend that you use
```python
pdf.render_to(pdfium.BitmapConv.numpy_ndarray, ...)
```
in favour of
```python
pdf.render_tonumpy(...)
```
because I intend to deprecate the aliases at some point.

**Comment by mara004** [comment]
I've released version 3 two days ago.
